### PR TITLE
Frictionless project

### PR DESCRIPTION
**Pull request by demenech**
**Ticket:** https://github.com/datopian/datahub-next/issues/76

Added some minor changes to this same PR so that we prevent conflicts

## Changes

- Minor
  - 404 page
    - Pull code previously merged into with-SSR that makes the 404-page work on the project page
  - Tests
    - Fixed tests that got broken after merging main into with-SSR (related to the trailing slash option, see notes below)
    - Created a 404-page test
    - Renamed __tests__/e2e/project-showcase.spec.js to __tests__/e2e/project.spec.js
  - Removed trailing slash. Found another way of doing the relative path by using a conditional `<base href=...` element
- Added new `datapackage` layout 
- `pages/@org/[org]/[project]/[...slug].tsx` renders general project files
- `pages/@org/[org]/[project]/index.tsx` renders a project page
- Created `lib/types.ts`
- `lib/project.ts` now exports a getProject function that returns a Project object that implements the ProjectProps interface

ss-2023-03-24-16:56:15

### CODE REVIEW

File: pages/@org/[org]/[project]/index.tsx [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+  if (project.metadata?.resources) {
+    frontMatter.layout = "datapackage";
+  }
+
+  return {
+    props: {
+      mdxSource,
+      frontMatter,
+      excerpt,
+      project: JSON.stringify(project),
```
Comment (rufuspollock) [inline pages/@org/[org]/[project]/index.tsx:62]: why are we json.stringify project? because it has dates etc in it?

**Review by rufuspollock — APPROVED** [review]
All looks good - left a few comments on style and approach. I think we can merge and rapidly iterate on the design.

File: lib/types.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
+export interface ProjectProps {
```
[the remark this answers is not in the document]
Comment (demenech) [reply lib/types.ts:1]: @rufuspollock yes. The reason is that we already have a class named Project, and we would have to use import aliases to be able to use both on the same file. IMO we don't need the interface in this case, just the class is enough, but depending on our intended design we might end up with interface and class names that are even more different.

File: lib/project.ts [hunk 3]
```diff
@@ -1,13 +1,47 @@
...
+  metadata: Metadata;
+
+  constructor(owner: string, name: string) {
+    this.name = name;
+    this.owner = owner;
+    this.github_repo = `https://github.com/${owner}/${name}`;
+  }
+
+  //  We can't use async in the constructor
+  init = async () => {
```
[the remark this answers is not in the document]
Comment (demenech) [reply lib/project.ts:18]: Noted. I wasn't thinking about the possibility of having two different "states", but this sounds good and it's a pattern similar to what dataview does with cached and non-cached files.

File: pages/@org/[org]/[project]/index.tsx (hunk 1 above)
Comment (demenech) [reply pages/@org/[org]/[project]/index.tsx:62]: In this case, I believe it's the methods, although we really need a serialization method seeing that metadata could assume different shapes (e.g could contain null values or dates).

I tested this and it works:

```typescript
const project2 = {
    name: project.name,
    metadata: project.metadata,
    owner: project.owner,
    readme: project.readme,
    github_repo: project.github_repo
  }

return {
  props: {
    mdxSource,
    frontMatter,
    excerpt,
    project: project2,
  },
```

We can have something like `Project.serialize()`.

**Pull request merged**